### PR TITLE
Improve solver factory

### PR DESCRIPTION
**Pull request by peterschrammel**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-checker/solver_factory.cpp [hunk 1]
```diff
@@ -28,6 +28,86 @@ Author: Daniel Kroening, Peter Schrammel
...
+  bool _output_xml_in_refinement)
+  : options(_options),
+    symbol_table(_symbol_table),
+    ns(_symbol_table),
+    message_handler(_message_handler),
+    output_xml_in_refinement(_output_xml_in_refinement)
+{
+}
+
+solver_factoryt::solvert::solvert()
```
Comment (tautschnig) [inline src/goto-checker/solver_factory.cpp]: Use `= default`?

File: src/goto-checker/solver_factory.cpp [hunk 2]
```diff
@@ -28,6 +28,86 @@ Author: Daniel Kroening, Peter Schrammel
...
+  if(options.get_bool_option("refine"))
+    return get_bv_refinement();
+  else if(options.get_bool_option("refine-strings"))
+    return get_string_refinement();
+  if(options.get_bool_option("smt2"))
+    return get_smt2(get_smt2_solver_type());
+  return get_default();
+}
+
+solver_factoryt::~solver_factoryt()
```
Comment (tautschnig) [inline src/goto-checker/solver_factory.cpp]: Use `= default`?

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ae245007cf30a750b0371e4abb70ca39b8cf2414).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95015757
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by smowton** [comment]
Why? A lot of these seem like the kind of one-liner ideally suited to inline definition

**Comment by tautschnig** [comment]
> Why? A lot of these seem like the kind of one-liner ideally suited to inline definition

1) I'm mainly worried about compilation time, and those "one-liners" require a lot of non-trivial header files.
2) Even if you just wanted to understand the interface, you are forced to read all the implementation.

To be honest, most of our inline definitions seem to be driven by laziness rather than conscious decisions about what to inline for performance or other data-driven reasons. And now that some cleanup is happening, we might as well take a look at those aspects as well. I'm ok with a rule like "you may define any function with up to N statements inline without providing performance data," but right now it seems entirely arbitrary.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 33daefed4054842e5fc57fc704c860cf26472a65).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95102698

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0471a5aa329dc8a4802822f642f2d76413529507).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95188405

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b76cd5fe6b4e7dca0da54857539a539ba70dac75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95419536

**Pull request merged**